Alerting does not work if record definitions are used in Alert rule

the tool does not trigger the alert if record definition is used in Alert rule. It shows empty result.

rules.yamls
```yaml
groups:
  - name: prometheus.rules
    rules:
      - record: httprequestratelow_record
        expr: http_requests{group="canary", job="app-server"} < 100 
      - alert: HTTPRequestRateLow
        expr: httprequestratelow_record
        for: 1m
        labels:
          severity: critical
```

test.yml
```yaml
name: Test HTTP Requests too low alert
rules:
  fromFile: rules.yaml
fixtures:
  - duration: 5m
    metrics:
      - http_requests{job="app-server", instance="0", group="canary", severity="overwrite-me"}	75 85  95 105 105  95  85
      - http_requests{job="app-server", instance="1", group="canary", severity="overwrite-me"}	80 90 100 110 120 130 140
assertions:
  - at: 0m
    expected:
      - alertname: HTTPRequestRateLow
        alertstate: pending
        group: canary
        instance: "0"
        job: app-server
        severity: critical
      - alertname: HTTPRequestRateLow
        alertstate: pending
        group: canary
        instance: "1"
        job: app-server
        severity: critical
    comment: |-
      At 0m, the alerts met the threshold but has not met the duration requirement. Expect the alert to be pending
```
pat test.yml
```
=== RUN   Test_HTTP_Requests_too_low_alert_at_0m
--- FAIL: Test_HTTP_Requests_too_low_alert_at_0m (0.00s)
	ruletest.go:91:
			Error Trace:	ruletest.go:91
			Error:      	Not equal:
			            	expected: 2
			            	actual  : 0
			Test:       	Test_HTTP_Requests_too_low_alert_at_0m
			Messages:   	Alert count does not match expected
	ruletest.go:212:
			Error Trace:	ruletest.go:212
			            				ruletest.go:92
			Error:      	Not equal:
			            	expected: []map[string]string{map[string]string{"group":"canary", "instance":"0", "job":"app-server", "severity":"critical", "__name__":"ALERTS", "alertname":"HTTPRequestRateLow", "alertstate":"pending"}, map[string]string{"alertstate":"pending", "group":"canary", "instance":"1", "job":"app-server", "severity":"critical", "__name__":"ALERTS", "alertname":"HTTPRequestRateLow"}}
			            	actual  : []map[string]string(nil)

			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -1,21 +1,2 @@
			            	-([]map[string]string) (len=2) {
			            	- (map[string]string) (len=7) {
			            	-  (string) (len=8) "__name__": (string) (len=6) "ALERTS",
			            	-  (string) (len=9) "alertname": (string) (len=18) "HTTPRequestRateLow",
			            	-  (string) (len=10) "alertstate": (string) (len=7) "pending",
			            	-  (string) (len=5) "group": (string) (len=6) "canary",
			            	-  (string) (len=8) "instance": (string) (len=1) "0",
			            	-  (string) (len=3) "job": (string) (len=10) "app-server",
			            	-  (string) (len=8) "severity": (string) (len=8) "critical"
			            	- },
			            	- (map[string]string) (len=7) {
			            	-  (string) (len=8) "__name__": (string) (len=6) "ALERTS",
			            	-  (string) (len=9) "alertname": (string) (len=18) "HTTPRequestRateLow",
			            	-  (string) (len=10) "alertstate": (string) (len=7) "pending",
			            	-  (string) (len=5) "group": (string) (len=6) "canary",
			            	-  (string) (len=8) "instance": (string) (len=1) "1",
			            	-  (string) (len=3) "job": (string) (len=10) "app-server",
			            	-  (string) (len=8) "severity": (string) (len=8) "critical"
			            	- }
			            	-}
			            	+([]map[string]string) <nil>

			Test:       	Test_HTTP_Requests_too_low_alert_at_0m
FAIL
```

--- a/pkg/bindata.go
+++ b/pkg/bindata.go
@@ -82,7 +82,7 @@ func schemaSchemaJson() (*asset, error) {
 		return nil, err
 	}
 
-	info := bindataFileInfo{name: "schema/schema.json", size: 1992, mode: os.FileMode(436), modTime: time.Unix(1530822196, 0)}
+	info := bindataFileInfo{name: "schema/schema.json", size: 1992, mode: os.FileMode(436), modTime: time.Unix(1530822892, 0)}
 	a := &asset{bytes: bytes, info: info}
 	return a, nil
 }

--- a/pkg/bindata.go
+++ b/pkg/bindata.go
@@ -82,7 +82,7 @@ func schemaSchemaJson() (*asset, error) {
 		return nil, err
 	}
 
-	info := bindataFileInfo{name: "schema/schema.json", size: 1992, mode: os.FileMode(420), modTime: time.Unix(1530413875, 0)}
+	info := bindataFileInfo{name: "schema/schema.json", size: 1992, mode: os.FileMode(436), modTime: time.Unix(1530822196, 0)}
 	a := &asset{bytes: bytes, info: info}
 	return a, nil
 }
@@ -102,7 +102,7 @@ func schemaSchemaYaml() (*asset, error) {
 		return nil, err
 	}
 
-	info := bindataFileInfo{name: "schema/schema.yaml", size: 1294, mode: os.FileMode(420), modTime: time.Unix(1529521862, 0)}
+	info := bindataFileInfo{name: "schema/schema.yaml", size: 1294, mode: os.FileMode(436), modTime: time.Unix(1530813958, 0)}
 	a := &asset{bytes: bytes, info: info}
 	return a, nil
 }

--- a/pkg/bindata.go
+++ b/pkg/bindata.go
@@ -82,7 +82,7 @@ func schemaSchemaJson() (*asset, error) {
 		return nil, err
 	}
 
-	info := bindataFileInfo{name: "schema/schema.json", size: 1992, mode: os.FileMode(436), modTime: time.Unix(1530822892, 0)}
+	info := bindataFileInfo{name: "schema/schema.json", size: 1992, mode: os.FileMode(436), modTime: time.Unix(1530823201, 0)}
 	a := &asset{bytes: bytes, info: info}
 	return a, nil
 }

--- a/pkg/ruleloader.go
+++ b/pkg/ruleloader.go
@@ -58,8 +58,9 @@ func (ruleLoader RuleLoader) Load() ([]*rules.Group, error) {
 				return nil, err
 			}
 
+			var rule rules.Rule
 			if r.Alert != "" {
-				rule := rules.NewAlertingRule(
+				rule = rules.NewAlertingRule(
 					r.Alert,
 					expr,
 					time.Duration(r.For),
@@ -67,8 +68,15 @@ func (ruleLoader RuleLoader) Load() ([]*rules.Group, error) {
 					labels.FromMap(r.Annotations),
 					logger,
 				)
-				rls = append(rls, rule)
 			}
+			if r.Record != "" {
+				rule = rules.NewRecordingRule(
+					r.Record,
+					expr,
+					labels.FromMap(r.Labels),
+				)
+			}
+			rls = append(rls, rule)
 		}
 		group := rules.NewGroup(rg.Name, filename, time.Duration(rg.Interval), rls, &rules.ManagerOptions{})
 		retval = append(retval, group)

--- a/pkg/ruleloader_test.go
+++ b/pkg/ruleloader_test.go
@@ -19,8 +19,9 @@ func TestRuleLoaderFromFile(t *testing.T) {
 	assert.Equal(t, 1, len(ruleGroups))
 	assert.Equal(t, "prometheus.rules", ruleGroups[0].Name())
 	assert.Equal(t, "testdata/rules.yaml", ruleGroups[0].File())
-	assert.Equal(t, 1, len(ruleGroups[0].Rules()))
+	assert.Equal(t, 2, len(ruleGroups[0].Rules()))
 	assert.Equal(t, "HTTPRequestRateLow", ruleGroups[0].Rules()[0].Name())
+	assert.Equal(t, "httprequestratelow_record", ruleGroups[0].Rules()[1].Name())
 }
 
 func TestRuleLoaderFromLiteral(t *testing.T) {
@@ -34,8 +35,9 @@ func TestRuleLoaderFromLiteral(t *testing.T) {
 	assert.Equal(t, 1, len(ruleGroups))
 	assert.Equal(t, "prometheus.rules", ruleGroups[0].Name())
 	assert.Equal(t, "__inline__", ruleGroups[0].File())
-	assert.Equal(t, 1, len(ruleGroups[0].Rules()))
+	assert.Equal(t, 2, len(ruleGroups[0].Rules()))
 	assert.Equal(t, "HTTPRequestRateLow", ruleGroups[0].Rules()[0].Name())
+	assert.Equal(t, "httprequestratelow_record", ruleGroups[0].Rules()[1].Name())
 }
 
 func TestRuleLoaderNoLoadingStrategySpecified(t *testing.T) {

--- a/pkg/ruletest.go
+++ b/pkg/ruletest.go
@@ -197,8 +197,10 @@ func alertLessThan(a, b map[string]string) bool {
 
 func assertAlertsEqual(t *testing.T, expected, actual []map[string]string) bool {
 	// Add __name__ attribute to the expected map
-	for _, e:= range expected {
-		e["__name__"] = "ALERTS"
+	for _, e := range expected {
+		if _, ok := e["__name__"]; !ok {
+			e["__name__"] = "ALERTS"
+		}
 	}
 
 	sort.SliceStable(expected, func(i, j int) bool {

--- a/pkg/ruletest_test.go
+++ b/pkg/ruletest_test.go
@@ -101,6 +101,15 @@ func TestAssertAlertsEqual(t *testing.T) {
 			},
 			comment: "Multiple alerts of the different order should also work",
 		},
+		{
+			expected: []Alert{
+				{"foo": "bar", "__name__": "superalert"},
+			},
+			actual: []map[string]string{
+				{"foo": "bar", "__name__": "superalert"},
+			},
+			comment: "If __name__ is specified by the user, do not override it",
+		},
 	}
 
 	for _, tc := range testCases {

--- a/pkg/testdata/rules.yaml
+++ b/pkg/testdata/rules.yaml
@@ -6,3 +6,5 @@ groups:
         for: 1m
         labels:
           severity: critical
+      - record: httprequestratelow_record
+        expr: http_requests{group="canary", job="app-server"} < 100

